### PR TITLE
EES-3583 remove all data sets option

### PR DESCRIPTION
**Pull request by amyb-hiveit**
- Adds a 'Remove all' button for data sets in the chart builder.
- Users are asked to confirm after clicking it.
- The data sets table and save actions are now always shown
- Trying to save without any data sets shows an error

datasets

### CODE REVIEW

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/ChartDataSetsConfiguration.tsx [hunk 1]
```diff
@@ -211,140 +214,168 @@ const ChartDataSetsConfiguration = ({
...
+          >
+            {buttons}
+          </ChartBuilderSaveActions>
+        </div>
+        {dataSets.length > 1 && (
+          <Button
+            className="dfe-align-self-end govuk-!-margin-left-2"
+            variant="secondary"
+            onClick={
+              isReordering ? toggleIsReordering.off : toggleIsReordering.on
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/ChartDataSetsConfiguration.tsx:360]: Think we could simplify this to: `onClick={toggleIsReordering}`

File: tests/robot-tests/tests/admin/bau/create_data_block_with_chart.robot [hunk 2]
```diff
@@ -337,6 +335,26 @@ Configure basic line chart
...
+    user checks chart legend item contains    id:chartBuilderPreview    1    Admission Numbers (Nailsea Youngwood)
+    user checks chart legend item contains    id:chartBuilderPreview    2    Admission Numbers (Syon)
+    user checks chart legend item contains    id:chartBuilderPreview    3    Admission Numbers (Bolton 001)
+
+Change data sets
+    user clicks link    Data sets
+
+    user clicks button    Remove all
+    user clicks button    Confirm
+
```
Comment (ntsim) [inline tests/robot-tests/tests/admin/bau/create_data_block_with_chart.robot:357]: This is a bit messy as the test cases should test specific things, but we're jamming a few things into `Configure basic line chart`.

Would suggest we organise the tests in the following way instead:

```robot
Configure basic line chart options
    user clicks button    Line
    user clicks radio    Set an alternative title
    user enters text into element    id:chartConfigurationForm-title    Test chart title
    user enters text into element    id:chartConfigurationForm-alt    Test chart alt
    user enters text into element    id:chartConfigurationForm-height    400
    user enters text into element    id:chartConfigurationForm-width    900

Validate changing data sets
    user clicks link    Data sets
    user waits until h3 is visible    Data sets
  
    user chooses select option    id:chartDataSetsConfigurationForm-location    Nailsea Youngwood
    user clicks button    Add data set
  
    user chooses select option    id:chartDataSetsConfigurationForm-location    Syon
    user clicks button    Add data set
  
    user chooses select option    id:chartDataSetsConfigurationForm-location    Bolton 001
    user clicks button    Add data set
  
    user checks chart legend item contains    id:chartBuilderPreview    1    Admission Numbers (Nailsea Youngwood)
    user checks chart legend item contains    id:chartBuilderPreview    2    Admission Numbers (Syon)
    user checks chart legend item contains    id:chartBuilderPreview    3    Admission Numbers (Bolton 001)

    user clicks button    Remove all
    user clicks button    Confirm

    # Note that we need some tests to check that the chart legend has been updated correctly as well!

Configure line chart data sets
    user chooses select option    id:chartDataSetsConfigurationForm-location    Nailsea Youngwood
    user clicks button    Add data set

    ...
```

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartDataSetsConfiguration.test.tsx [hunk 3]
```diff
@@ -978,5 +995,44 @@ describe('ChartDataSetsConfiguration', () => {
...
+      userEvent.click(screen.getByRole('button', { name: 'Remove all' }));
+
+      const modal = within(screen.getByRole('dialog'));
+      expect(modal.getByText('Remove all data sets')).toBeInTheDocument();
+      expect(
+        modal.getByText('Are you sure you want to remove all data sets?'),
+      ).toBeInTheDocument();
+
+      userEvent.click(modal.getByRole('button', { name: 'Confirm' }));
+
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartDataSetsConfiguration.test.tsx:1034]: We haven't actually check that the data sets have been removed

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/ChartDataSetsConfiguration.tsx [hunk 4]
```diff
@@ -211,140 +214,168 @@ const ChartDataSetsConfiguration = ({
...
+      >
+        <Droppable droppableId="dataSets" isDropDisabled={!isReordering}>
+          {(droppableProvided, droppableSnapshot) => (
+            <table>
+              <thead>
+                <tr>
+                  <th>Data set</th>
+                  {!isReordering && (
+                    <th className="dfe-align--right">
+                      <VisuallyHidden>Remove data sets:</VisuallyHidden>
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/ChartDataSetsConfiguration.tsx:258]: This doesn't look right, shouldn't we do something like this instead?
```tsx
<ButtonText>
  Remove all <VisuallyHidden> data sets</VisuallyHidden>
</ButtonText>
```

**Review by ntsim — CHANGES_REQUESTED** [review]

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartDataSetsConfiguration.test.tsx (hunk 3 above)
Comment (amyb-hiveit) [reply src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartDataSetsConfiguration.test.tsx:1036]: I don't think it's possible in these tests as it's handled by ChartBuilder which then passes the datasets down to this component. I've added some tests for this for ChartBuilder, and a couple of other simple ones. I think most of the chart tabs functionality is tested in the individual tab tests but there may be more tests we want to add here in future.

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartBuilder.test.tsx [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+    dataSets: {
+      isValid: true,
+      submitCount: 0,
+      id: 'dataSets',
+      title: 'Data sets',
+    },
+  };
+
+  const testTable = mapFullTable(testTableData);
+  const testMeta: TableDataSubjectMeta = {
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartBuilder.test.tsx:33]: This doesn't seem right - we're calling `mapFullTable` on some random other test data here when we're defining some suite-specific `testMeta` / `testSubjectMeta` that may not have any correlation to `testTableData`.

Would suggest something like the following instead:

```tsx
const testTableData: TableDataResponse = {
  subjectMeta: { ... },
  results: [ ... ]
};

const testTable = mapFullTable(testTableData);

<ChartBuilder
  results={testTable.results}
  meta={testTable.subjectMeta}
/>
```

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartDataSetsConfiguration.test.tsx (hunk 3 above)
Comment (ntsim) [reply src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartDataSetsConfiguration.test.tsx:1036]: Ah good point! See that you've added some tests for `ChartBuilder` to test this though, nice one! :+1:

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartBuilder.test.tsx [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+
+  test('renders the chart type selector only when no initial value', () => {
+    render(
+      <ChartBuilderFormsContextProvider initialForms={testFormState}>
+        <ChartBuilder
+          releaseId="release-1"
+          data={testTable.results}
+          meta={testSubjectMeta}
+          tableTitle="Table title"
+          onChartSave={jest.fn(() => Promise.resolve())}
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartBuilder.test.tsx:108]: Bit of a nitpick, but couldn't we just pass in `jest.fn()` for all `onChartSave` and `onTableQueryUpdate` props?

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartBuilder.test.tsx [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+      screen.getByRole('tab', { name: 'Chart configuration' }),
+    ).toBeInTheDocument();
+    expect(screen.getByRole('tab', { name: 'Data sets' })).toBeInTheDocument();
+    expect(screen.getByRole('tab', { name: 'Legend' })).toBeInTheDocument();
+    expect(
+      screen.getByRole('tab', { name: 'X Axis (major axis)' }),
+    ).toBeInTheDocument();
+    expect(
+      screen.getByRole('tab', { name: 'Y Axis (minor axis)' }),
+    ).toBeInTheDocument();
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartBuilder.test.tsx:167]: Could improve the testing of this by checking all of the tabs:

```ts
const tabs = screen.getAllByRole('tab');

expect(tabs).toHaveLength(5);

expect(tabs[0]).toHaveTextContent('Chart configuration');
expect(tabs[0]).toHaveAttribute('aria-selected', 'true');

// and so on...
```

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartBuilder.test.tsx [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+      });
+
+      const updatedTableRows = screen.getAllByRole('row');
+      expect(updatedTableRows).toHaveLength(3);
+      expect(updatedTableRows[2]).toHaveTextContent(
+        'Number of unauthorised absence sessions (Female, All locations, All time periods)',
+      );
+    });
+
+    test('removing data sets', async () => {
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartBuilder.test.tsx:238]: Would split out two test cases to make this a little cleaner - one for removing a single data set and one for removing all data sets.

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartBuilder.test.tsx [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+      );
+
+      await waitFor(() => {
+        expect(
+          screen.queryByText(
+            'Number of unauthorised absence sessions (Male, All locations, All time periods)',
+          ),
+        ).not.toBeInTheDocument();
+      });
+
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartBuilder.test.tsx:293]: We should check that there are still 3 data sets remaining to ensure there are no bugs that might remove more than one.

File: tests/robot-tests/tests/admin/bau/create_data_block_with_chart.robot [hunk 10]
```diff
@@ -335,11 +333,34 @@ Configure basic line chart
...
+    user clicks button    Add data set
+
+    user checks chart legend item contains    id:chartBuilderPreview    1    Admission Numbers (Nailsea Youngwood)
+    user checks chart legend item contains    id:chartBuilderPreview    2    Admission Numbers (Syon)
+    user checks chart legend item contains    id:chartBuilderPreview    3    Admission Numbers (Bolton 001)
+
+    user clicks button    Remove all
+    user clicks button    Confirm
+
+    user checks page contains    Configure the chart and add data to view a preview
```
Comment (ntsim) [inline tests/robot-tests/tests/admin/bau/create_data_block_with_chart.robot:356]: Would go a little further here and check that the data sets table has been emptied too.

**Review by ntsim — CHANGES_REQUESTED** [review]

File: src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartBuilder.test.tsx (hunk 5 above)
Comment (amyb-hiveit) [reply src/explore-education-statistics-admin/src/pages/release/datablocks/components/chart/__tests__/ChartBuilder.test.tsx]: yeah don't know what I was doing there! Sorted now.

**Review by ntsim — APPROVED** [review]

**Pull request merged**